tests/specs/mcd/verification.k

```diff
@@ -130,6 +125,33 @@ module LEMMAS-MCD-HASKELL [symbolic, kore]
        andBool Y <=Int 0
        andBool notBool #rangeUInt(256, X -Int Y) [simplification]
 
+    rule { chop(X) #Equals chop(Y) } => { X #Equals Y }
+         requires #rangeSInt(256, X) andBool #rangeSInt(256, Y)
+         [simplification]
+
+    rule { chop ( X:Int ) #Equals chop ( Y:Int ) } =>
+         { true #Equals X ==Int Y andBool Y <=Int 0 }
+         requires (minSInt256 <=Int X) andBool (X <=Int 0) andBool #rangeSInt(256, Y) [simplification]
+
+  // ########################
+  // Buffer Reasoning
+  // ########################
+
+    rule { BA1:ByteArray ++ BA2:ByteArray #Equals BA1:ByteArray ++ BA3:ByteArray } => { BA2 #Equals BA3 } [simplification]
+
+    rule { BA1:ByteArray ++ (BA2:ByteArray ++ BA3:ByteArray) #Equals BA4:ByteArray ++ BA3 } => { (BA1 ++ BA2) #Equals BA4 } [simplification]
+
+    rule { #buf(32, BY:Int):ByteArray #Equals #buf(32, BZ:Int):ByteArray } => { BY #Equals BZ }
+         requires #rangeUInt(256, BY) andBool #rangeUInt(256, BZ) [simplification]
+
+    rule { #buf(32, BY:Int):ByteArray #Equals #buf(32, BZ:Int):ByteArray ++ BA2:ByteArray } =>
+         { true #Equals (BY ==Int BZ) andBool (.ByteArray ==K BA2) }
+         requires #rangeUInt(256, BY) andBool #rangeUInt(256, BZ) [simplification]
+
+    rule { #buf(32, BY:Int):ByteArray ++ BA1:ByteArray #Equals #buf(32, BZ:Int):ByteArray ++ BA2:ByteArray } =>
+         { true #Equals (BY ==Int BZ) andBool (BA1 ==K BA2) }
+         requires #rangeUInt(256, BY) andBool #rangeUInt(256, BZ) [simplification]
```

Comment:
We can probably make this rule simpler and more general:

```suggestion
    rule { BA1 ++ BA2:ByteArray #Equals BA3:ByteArray ++ BA4:ByteArray } => #Top
         requires BA1 ==K BA3 andBool BA2 ==K BA4 [simplification]
```

Then the other rules can take care of the simplification of the side-conditions.

Reply:
This simplification and the one above are applied in the context in which we know the LHS holds, it is in the positive constraints of the example. It is a straightforward inference that does not branch into further `requires` checks. The `requires` of the suggested simplification will be checked for every concatenation equality. Moreover, the goal is not to discharge the LHS into `#Top`, the goal is to extract more information, so I would like to keep the simplification as-is. Having said that, it is an instance of a more general simplification:

```
rule { BA1 ++ BA2:ByteArray #Equals BA3:ByteArray ++ BA4:ByteArray } => { true #Equals BA1 ==K BA3 andBool BA2 ==K BA4 } 
requires #length(BA1) ==K #length(BA3)
```

where `#length` returns the buffer length (I don't think we have that?), which, in this case, is 32.

Reply:
We do have `#length`, it's called `#sizeByteArray`. I like your suggested approach here ^^^ better than the original, in general I like more general lemmas better, they are easier to show correct and allow decomposition of reasoning. Specifically, it's nice that the new lemma does not mention `#bufStrict` symbol at all. Maybe we want to remove or replace that construct eventually? Doing so would be easier if fewer lemmas mentioned it directly.

Reply:
```
    rule { BA1:ByteArray ++ BA2:ByteArray #Equals BA3:ByteArray ++ BA4:ByteArray } =>
         { true #Equals BA1 ==K BA3 andBool BA2 ==K BA4 }
         requires (#sizeByteArray(BA1) ==Int #sizeByteArray(BA3))
           orBool (#sizeByteArray(BA2) ==Int #sizeByteArray(BA4))
         [simplification]
```

and

```
    rule { BA1:ByteArray #Equals BA2:ByteArray ++ BA3:ByteArray } =>
         { true #Equals BA1 ==K BA2 andBool .ByteArray ==K BA3 }
         requires (#sizeByteArray(BA1) ==Int #sizeByteArray(BA2))
         [simplification]
```

appear to do the trick.